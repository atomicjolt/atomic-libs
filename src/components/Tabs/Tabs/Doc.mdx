import {
  Title,
  Subtitle,
  Description,
  Primary,
  ArgsTable,
  Stories,
} from "@storybook/addon-docs";

<Title />
<Subtitle />
<Description />
<Primary />
<ArgsTable />
<Stories />

# Uncontrolled vs Controlled Tabs

By default, the `<Tab />` component controlls the current tab internally.

But you can take control of the this using the `currentTab` and `onChange` props.
This is useful for integrating the Tab component into things like react-router.

```tsx
import { useMemo } from "react";
import { useLocation, useHistory } from "react-router-dom";
import { Tabs, Tab } from "atomic-hooks";

const ROUTES = ["/preview", "/edit", "/reports"];

function ControlledTabs() {
  const history = useHistory();
  const { pathname } = useLocation();

  const currentTab = useMemo(() => ROUTES.indexOf(pathname), [pathname]);

  const onChange = (idx: number) => {
    history.push(ROUTES[idx]);
  };

  return (
    <Tabs currentTab={currentTab} onChange={onChange}>
      <Tab label="Tab One">
        <div> I'm the first tab</div>
      </Tab>

      <Tab label="Tab Two">
        <div> I'm the second tab</div>
      </Tab>
    </Tabs>
  );
}
```
