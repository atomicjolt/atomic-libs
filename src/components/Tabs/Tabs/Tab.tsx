import React from "react";
import { useTabContext } from "./utils";

interface TabProps {
  children: React.ReactNode;
  /** The name of tab, when `currentTab` matches this name, this tab will be rendered */
  name: string;
}

export default function Tab({ children, name }: TabProps) {
  const { currentTab } = useTabContext();

  if (name !== currentTab) return null;

  return <>{children}</>;
}

Tab.displayName = "Tabs.Tab";
