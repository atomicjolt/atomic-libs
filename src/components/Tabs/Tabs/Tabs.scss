@use "../../mixins";

.aje-tab-list {
  --tab-border-width: 1px;
  --tab-border-clr: var(--border-clr-primary);
  --tab-padding-hori: 16px;
  --tab-padding-vert: 8px;

  margin-bottom: calc(var(--tab-border-width) * -1);
  display: flex;
  align-items: center;
  gap: 2px;
}

.aje-tab {
  --tab-radius: var(--radius);
  --tab-bg-clr: var(--neutral100);
  --tab-border-width: 1px;
  --tab-border-clr: transparent;
  --tab-text-clr: var(--text-clr-alt);

  display: inline-block;
  padding: var(--tab-padding-vert) var(--tab-padding-hori);
  background-color: var(--tab-bg-clr);
  border-radius: var(--tab-radius) var(--tab-radius) 0 0;
  border: var(--tab-border-width) solid;
  border-color: var(--tab-border-clr);
  position: relative;
  @include mixins.regular;
  font-size: 1.6rem;
  text-decoration: none;
  color: var(--tab-text-clr);

  &[aria-selected="true"] {
    --tab-bg-clr: var(--neutral50);
    --tab-border-clr: var(--border-clr-primary);
    --tab-text-clr: var(--text-clr);

    &::after {
      content: "";
      position: absolute;
      height: calc(var(--tab-border-width) + 2px);
      bottom: -1px;
      left: 0;
      right: 0;
      background-color: var(--tab-bg-clr);
      z-index: 3;
    }
  }

  &:hover {
    --tab-text-clr: var(--text-clr);
    --tab-border-clr: var(--border-clr-primary);
    cursor: pointer;
  }
  &:focus-visible {
    outline: var(--outline);
    outline-offset: -1px;
  }
}

.aje-tab-content {
  --tab-bg-clr: var(--neutral50);
  --tab-border-clr: var(--border-clr-primary);

  position: relative;
  z-index: 2;
  background-color: var(--tab-bg-clr);
  border-top: 1px solid var(--tab-border-clr);
  min-height: 4px;

  &:focus {
    outline: none;
  }
  &:focus-visible {
    outline: var(--outline);
    outline-offset: -2px;
  }
}

.aje-tab-link {
  position: relative;
}

.aje-tab-notification {
  --tab-notif-error-clr: var(--error600);
  --tab-notif-info-clr: var(--primary400);

  position: absolute;
  width: 10px;
  height: 10px;
  line-height: 10px;
  border-radius: 50%;
  text-align: center;
  top: var(--tab-padding-vert);
  right: calc(var(--tab-padding-hori) - 8px);

  &--error {
    background-color: var(--tab-notif-error-clr);
  }

  &--info {
    background-color: var(--tab-notif-into-clr);
  }
}
