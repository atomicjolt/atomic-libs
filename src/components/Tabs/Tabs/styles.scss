@use "../../defines.scss" as *;

.aj-tab-list {
  --tab-border-width: 1px;
  --tab-border-clr: var(--border-clr-primary);

  margin-bottom: calc(var(--tab-border-width) * -1);
  display: flex;
  align-items: center;
  gap: 2px;
}

.aj-tab {
  --tab-radius: var(--radius);
  --tab-padding-hori: 16px;
  --tab-padding-vert: 8px;
  --tab-bg-clr: var(--neutral100);
  --tab-border-width: 1px;
  --tab-border-clr: transparent;
  --tab-text-clr: var(--text-clr-alt);

  display: inline-block;
  padding: var(--tab-padding-vert) var(--tab-padding-hori);
  background-color: var(--tab-bg-clr);
  border-radius: var(--tab-radius) var(--tab-radius) 0 0;
  border: var(--tab-border-width) solid;
  border-color: var(--tab-border-clr);
  position: relative;
  @include regular;
  font-size: 1.6rem;
  text-decoration: none;
  color: var(--tab-text-clr);

  &[aria-selected="true"] {
    --tab-bg-clr: var(--neutral50);
    --tab-border-clr: var(--border-clr-primary);
    --tab-text-clr: var(--text-clr);

    &::after {
      content: "";
      position: absolute;
      height: calc(var(--tab-border-width) + 2px);
      bottom: -1px;
      left: 0;
      right: 0;
      background-color: var(--tab-bg-clr);
      z-index: 3;
    }
  }

  &:hover {
    --tab-text-clr: var(--text-clr);
    --tab-border-clr: var(--border-clr-primary);
  }
  &:focus-visible {
    outline: var(--outline);
    outline-offset: -1px;
  }
}

.aj-tab-content {
  --tab-bg-clr: var(--neutral50);
  --tab-border-clr: var(--border-clr-primary);

  position: relative;
  z-index: 2;
  background-color: var(--tab-bg-clr);
  border-top: 1px solid var(--tab-border-clr);
  min-height: 4px;

  &:focus {
    outline: none;
  }
  &:focus-visible {
    outline: var(--outline);
    outline-offset: -2px;
  }
}
