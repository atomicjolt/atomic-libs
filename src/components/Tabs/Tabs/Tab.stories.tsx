import { ComponentMeta, ComponentStory } from "@storybook/react";
import Tabs, { Tab, Props } from ".";
import Doc from "./Doc.mdx";

export default {
  title: "Tabs/Tabs",
  component: Tabs,
  parameters: {
    docs: {
      page: Doc,
    },
  },
} as ComponentMeta<typeof Tabs>;

const Template: ComponentStory<typeof Tabs> = (args: Props) => (
  <Tabs {...args} />
);

export const Default = Template.bind({});
Default.args = {
  tabs: { tab1: { label: "Tab One", notification: "error" }, tab2: "Tab Two" },
  children: [
    <Tab name="tab1" key="key1">
      <div style={{ paddingTop: "3em" }}>
        <h2>First Tab</h2>
      </div>
    </Tab>,
    <Tab name="tab2" key="key2">
      <div style={{ paddingTop: "3em" }}>
        <h2>Second Tab</h2>
      </div>
    </Tab>,
  ],
};

// Additional stories go here
