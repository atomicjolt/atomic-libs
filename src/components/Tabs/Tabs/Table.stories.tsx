import { ComponentMeta, ComponentStory } from "@storybook/react";
import Tabs, { Props } from ".";

export default {
title: "Tabs/Tabs",
component: Tabs
} as ComponentMeta<typeof Tabs>

const Template: ComponentStory<typeof Tabs> = (args: Props) => (
  <Tabs {...args} />
);

export const Default = Template.bind({});
Default.args = {
  label: "Tab Label",
  linkTarget: "https://www.atomicjolt.com/",
}

// Additional stories go here