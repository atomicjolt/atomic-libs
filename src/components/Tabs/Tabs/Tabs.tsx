import React from "react";
import cn from "classnames";
import styled from "styled-components";
import { useIds } from "../../../hooks";
import { HasChildren, HasClassName } from "../../../types";
import Tab from "./Tab";
import TabsNavigation, { TabLink } from "./TabNavigation";
import { TabsContextT } from "./Tabs.types";
import { getTabId, TabsContext, useTabContext } from "./utils";

const TabContentWrapper = styled.div`
  --tab-bg-clr: var(--neutral50);
  --tab-border-clr: var(--border-clr-primary);

  position: relative;
  z-index: 2;
  background-color: var(--tab-bg-clr);
  border-top: 1px solid var(--tab-border-clr);
  min-height: 4px;

  &:focus {
    outline: none;
  }
  &:focus-visible {
    outline: var(--outline);
    outline-offset: -2px;
  }
`;

export interface TabsProps<T extends string> extends HasChildren {
  readonly currentTab: T;
  readonly onChange: (name: T) => void;
}

function Tabs<T extends string>(props: TabsProps<T>) {
  const { children, currentTab, onChange } = props;
  const [tabIdPrefix, tabContentId] = useIds("Tabs", ["prefix", "content"]);

  const ctx: TabsContextT<T> = {
    changeTab: (name: T) => onChange && onChange(name),
    currentTab: currentTab,
    tabContentId,
    tabIdPrefix,
  };

  return <TabsContext.Provider value={ctx}>{children}</TabsContext.Provider>;
}

type TabsContentProps = HasChildren & HasClassName;

export function TabsContent(props: TabsContentProps) {
  const { children, className } = props;
  const { tabContentId, tabIdPrefix, currentTab } = useTabContext();

  return (
    <TabContentWrapper
      className={cn("aje-tab-content", className)}
      id={tabContentId}
      tabIndex={0}
      role="tabpanel"
      aria-labelledby={getTabId(tabIdPrefix, currentTab)}
    >
      {children}
    </TabContentWrapper>
  );
}

TabsContent.displayName = "Tabs.Content";

// @ts-ignore
Tabs.Tab = Tab;
// @ts-ignore
Tabs.Navigation = TabsNavigation;
// @ts-ignore
Tabs.Link = TabLink;
// @ts-ignore
Tabs.Content = TabsContent;

export default Tabs;
