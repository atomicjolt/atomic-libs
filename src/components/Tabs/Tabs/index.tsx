import React from 'react'
import '../../general.scss'
import './styles.scss'

export interface Props {
  /** Must include a label for the tab. */
  label: string;
  linkTarget: string;
  content?: any;
}

/** Tabs Component */
export default function Tabs({ 
    label,
    linkTarget,
    content,
  }: Props) {

  const tabID = label + 'Tab';
  const tabContentID = label + 'Content';
  const selected = false;
  const current = selected ? 'page' : undefined;

  return (
    <>
      <div className="aj-tab-list" role="tablist" aria-label="navigation">
        <a
          className="aj-tab"
          id={tabID}
          role="tab"
          aria-controls={tabContentID}
          aria-selected="true"
          aria-current={current}
          href={linkTarget}
        >
          {label}
        </a>
        <a
          className="aj-tab"
          id={tabID}
          role="tab"
          aria-controls={tabContentID}
          aria-selected={selected}
          aria-current={current}
          href={linkTarget}
          tabIndex={-1}
        >
          {label}
        </a>
      </div>
      <div
        className="aj-tab-content"
        id={tabContentID}
        tabIndex={0}
        role="tabpanel"
        aria-labelledby={tabID}
      >
        {content}
      </div>
    </>
  )
}
