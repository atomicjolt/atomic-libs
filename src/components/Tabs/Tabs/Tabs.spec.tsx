import React from "react";
import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import Tab from "./Tab";
import TabsNavigation, { TabLink } from "./TabNavigation";
import Tabs, { TabsContent } from "./Tabs";

test("matches snapshot", () => {
  const res = render(
    <Tabs currentTab="tab1" onChange={() => {}}>
      <TabsNavigation key="nav">
        <TabLink to="tab1">Tab One</TabLink>
        <TabLink to="tab2">Tab Two</TabLink>
      </TabsNavigation>
      <TabsContent key="content">
        <Tab name="tab1">
          <div>
            <h2>First Tab</h2>
          </div>
        </Tab>
        <Tab name="tab2">
          <div>
            <h2>Second Tab</h2>
          </div>
        </Tab>
      </TabsContent>
    </Tabs>
  );

  expect(res.asFragment()).toMatchSnapshot();
});
