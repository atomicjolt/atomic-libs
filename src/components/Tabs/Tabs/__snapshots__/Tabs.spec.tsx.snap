// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="navigation"
    class="aje-tab-list"
    role="tablist"
  >
    <div
      class="aje-tab-link"
    >
      <a
        aria-controls="Tabs-:r0:-content"
        aria-current="true"
        aria-selected="true"
        class="aje-tab"
        id="Tabs-:r0:-prefix-tab1"
        role="tab"
        tabindex="-1"
      >
        Tab One
      </a>
    </div>
    <div
      class="aje-tab-link"
    >
      <a
        aria-controls="Tabs-:r0:-content"
        aria-current="false"
        aria-selected="false"
        class="aje-tab"
        id="Tabs-:r0:-prefix-tab2"
        role="tab"
        tabindex="-1"
      >
        Tab Two
      </a>
    </div>
  </div>
  <div
    aria-labelledby="Tabs-:r0:-prefix-tab1"
    class="aje-tab-content"
    id="Tabs-:r0:-content"
    role="tabpanel"
    tabindex="0"
  >
    <div>
      <h2>
        First Tab
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
